python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
+        Returns
+        -------
+        BasicResultIndicator
+            An indicator that, when successful, includes as ``data`` the serialized status string of the provided job.
+        """
+        # TODO: implement
+        raise NotImplementedError('{} function "request_job_status" not implemented yet'.format(self.__class__.__name__))
+
+    async def request_job_stop(self, job_id: str, *args, **kwargs) -> ResultIndicator:
         """
-        Download an entire dataset to a local directory.
+        Request the provided job be stopped; i.e., transitioned to the ``STOPPED`` exec step.
 
         Parameters
         ----------
-        dataset_name : str
-            The dataset of interest.
-        dest_dir : Path
-            Path to the local directory to which to save the dataset's data.
+        job_id : str
+            The id of the job in question.
+        args
+            (Unused) variable positional args.
+        kwargs
+            (Unused) variable keyword args.
 
         Returns
         -------
-        bool
-            Whether the download was successful.
+        ResultIndicator
+            An indicator of whether the job was stopped as requested.
         """
-        pass
+        # TODO: implement
+        raise NotImplementedError('{} function "request_job_stop" not implemented yet'.format(self.__class__.__name__))
 
-    @abstractmethod
-    async def download_from_dataset(self, dataset_name: str, item_name: str, dest: Path) -> bool:
+    async def request_jobs_list(self, jobs_list_active_only: bool, *args, **kwargs) -> BasicResultIndicator:
         """
-        Download a specific item within a dataset to a local path.
+        Request a list of ids of existing jobs.
 
-        Exactly what an "item" is is implementation specific, and should be documented.
+        The list of ids will be included as the ``data`` attribute of the returned ::class:`ResultIndicator`.
 
         Parameters
         ----------
-        dataset_name : str
-            The dataset of interest.
-        item_name : str
-            The name of the item within a dataset to download.
-        dest : Path
-            A local path at which to save the downloaded item.
+        jobs_list_active_only : bool
+            Whether to exclusively include jobs with "active" status values.
+        args
+            (Unused) variable positional args.
+        kwargs
+            (Unused) variable keyword args.
 
         Returns
         -------
+        BasicResultIndicator
+            An indicator that, when successful, includes as ``data`` the list of ids of existing jobs.
+
+        See Also
+        -------
+        get_jobs_list
+        """
+        # TODO: implement
+        raise NotImplementedError('{} function "request_jobs_list" not implemented yet'.format(self.__class__.__name__))
+
+    async def submit_ngen_request(self, **kwargs) -> NGENRequestResponse:
```

Comment:
@aaraney, the main reason was that - thinking from the perspective of CLI usage - I didn't want _DmodClient_ to have to be responsible for message initialization or deserialization.  The specialized client needed to handle deserialization.  Given the usage setup, this required it to be capable of just accepting passed-through CLI parsed args (and required `__main__` to be somewhat particular about the names of the parsed args it allowed).

That said, it does seem as though either the functions are misnamed or they need to _also_ accommodate complete objects.  Would it be too messy if, say, I checked for the presence of a `request` key in `kwargs` and used that directly if it was request object?

Reply:
Pinging on this again.

Reply:
> the main reason was that - thinking from the perspective of CLI usage - I didn't want _DmodClient_ to have to be responsible for message initialization or deserialization.

Yeah I hear that argument and I agree. I don't think either `DmodClient` nor the `JobClient` should be responsible for message initialization or deserialization. After taking a step back, I think this boils down to a difference in option of when to use factory functions. `DmodClient.execute_job` is a great example of this. Below is the signature for convenience sake:

```python
    async def execute_job(self, workflow: str, **kwargs) -> ResultIndicator:
```

IMO I would have implemented it as something like:

```
    async def execute_job(self, job: ExternalAbstractNgenRequest) -> ResultIndicator:
```

Right the burden of initialization or deserialization is now put on whoever uses a `DmodClient`, but IMO that is a good thing. If the code that _uses_ a `DmodClient` needs to do some custom logic to transform a request or deserialize the request from different sources, that is up to the code that _uses_ the client. It also simplifies how code that _uses_ a `DmodClient` handle failures. With the current implementation an `execute_job` can fail for numerous reasons. try catching would be a little painful to disseminate good error messages back to a user. I think we can simplify the `DmodClient` and `JobClient` by shifting the burden of initialization / deserialization to the code that _uses_ a `DmodClient`. 

You know me, I am a big fan of dependency injection and sometimes overly so. You have a lot more experience developing software than I do and I really value you thoughts. So I dont mean to drag this on, but it would be really valuable for me to hear your argument for using a factory here.

Im happy to keep it as is, but I do hope to learn something from our discussion!

Reply:
> I don't think either DmodClient nor the JobClient should be responsible for message initialization or deserialization. .... Right the burden of initialization or deserialization is now put on whoever uses a DmodClient, but IMO that is a good thing.

I'm not sure that makes sense.  I think of _DmodClient_ as intended for use by something outside of DMOD (for purposes of this discussion, I think it reasonable to treat the functions handling interfacing with the command line as on the "outside").   We could still put the (application logic) burden on them, but then what purpose do these client classes serve?  And would it not be better to provide something that encapsulates this behavior (if so, does that belong somewhere else)?

Reply:
> We could still put the (application logic) burden on them, but then what purpose do these client classes serve? And would it not be better to provide something that encapsulates this behavior (if so, does that belong somewhere else)?

I see your point. I've been thinking of these client classes more like an RPC client where the main purposes of the client are to handle connection, session state, transmission, and reception of data. Not like an HTTP request client where the underlying protocol is flat in nature and easier to abstract. In either case, with the existing API there is nothing stopping you from creating a request object (pydantic model), dumping that to a dictionary, and passing that in as input. I assumed building up a request would be the primary way to interface with the api (or deserializing user input and then passing in the object).

The one issue with passing in a request object is mutating / attaching session information. We've run into this in other contexts before. I am not sure if accepting `**kwargs` makes that any easier or has any benefits.